### PR TITLE
enh (media): Add migration script

### PR DESCRIPTION
**Pull request by callapa**
## Description

Add a migration script to migrate media from an on-premise platform to a Cloud platform.

**Fixes** MON-24294

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [x] 24.04.x (master)

#### Community contributors & Centreon team

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/packaging/centreon.spectemplate
Comment (kduret) [inline centreon/packaging/centreon.spectemplate:1]: missing debian packaging, update should be probably here : https://github.com/centreon/centreon/blob/develop/centreon/packaging/debian/centreon-web.install

File: centreon/src/Core/Application/Common/UseCase/CliAbstractPresenter.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
+<?php
```
Comment (a-launois) [inline centreon/src/Core/Application/Common/UseCase/CliAbstractPresenter.php:1]: Why is it in Core/Application/Common and not in Core/Common/Application ?

File: centreon/src/Core/Media/Infrastructure/Repository/ApiWriteMediaRepository.php [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        $options['headers'] = $formData->getPreparedHeaders()->toArray();
+        $options['headers'][] = 'X-AUTH-TOKEN: ' . $this->authenticationToken;
+        $options['body'] = $formData->bodyToString();
+
+        $response = $this->httpClient->request('POST', $apiEndpoint, $options);
+        if ($response->getStatusCode() !== 200) {
+            /**
+             * @var array{message: string} $content
+             */
+            $content = json_decode($response->getContent(), true);
```
Comment (a-launois) [inline centreon/src/Core/Media/Infrastructure/Repository/ApiWriteMediaRepository.php:112]: I think getContent() throws an error in this case
I used toArray(true) instead

Comment (callapa) [reply centreon/src/Core/Media/Infrastructure/Repository/ApiWriteMediaRepository.php:112]: > I think getContent() throws an error in this case

Not every time, but I'll change it

File: centreon/src/Core/Application/Common/UseCase/CliAbstractPresenter.php (hunk 1 above)
Comment (callapa) [reply centreon/src/Core/Application/Common/UseCase/CliAbstractPresenter.php:1]: To put it in the same place as the AbstractPresenter

File: centreon/src/Core/Media/Infrastructure/Command/MigrateAllMedias/MigrateAllMediasCommand.php [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class MigrateAllMediasCommand extends AbstractMigrationCommand
+{
+    use LoggerTrait;
+
+    protected static $defaultName = 'media:all';
+
+    protected static $defaultDescription = 'Migrate all medias from the current platform to the defined target platform';
```
Comment (cg-tw) [inline centreon/src/Core/Media/Infrastructure/Command/MigrateAllMedias/MigrateAllMediasCommand.php:42]:
```suggestion
    protected static $defaultDescription = 'Migrate all media from the current platform to the defined target platform';
```

File: centreon/src/Core/Media/Infrastructure/Command/MigrateAllMedias/MigrateAllMediasCommand.php [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        $this->maxFilesize = $this->parseSize($maxFilesize);
+        $this->postMax = $this->parseSize($postMax);
+    }
+
+    protected function configure(): void
+    {
+        $this->addArgument(
+            'target-url',
+            InputArgument::REQUIRED,
+            "The target platform base url to connect to the API (ex: 'http://localhost/centreon')"
```
Comment (cg-tw) [inline centreon/src/Core/Media/Infrastructure/Command/MigrateAllMedias/MigrateAllMediasCommand.php:66]:
```suggestion
            "The target platform base URL to connect to the API (ex: 'http://localhost/centreon')"
```

File: centreon/src/Core/Media/Infrastructure/Repository/DbReadMediaRepository.php [hunk 5]
```diff
@@ -63,4 +75,189 @@ public function existsByPath(string $path): bool
...
+
+    /**
+     * To avoid loading all database elements at once, this iterator allows you to retrieve them in blocks of
+     * MAX_ITEMS_BY_REQUEST elements.
+     *
+     * @return \Iterator<int, Media>&\Countable
+     */
+    private function findAllByIteration(): \Iterator&\Countable
+    {
+        $this->logger->debug(sprintf('Loading medias in blocks of %d elements', self::MAX_ITEMS_BY_REQUEST));
```
Comment (cg-tw) [inline centreon/src/Core/Media/Infrastructure/Repository/DbReadMediaRepository.php:95]:
```suggestion
        $this->logger->debug(sprintf('Loading media in blocks of %d elements', self::MAX_ITEMS_BY_REQUEST));
```

File: centreon/src/Core/Media/Infrastructure/Repository/DbReadMediaRepository.php [hunk 6]
```diff
@@ -63,4 +75,189 @@ public function existsByPath(string $path): bool
...
+                }
+
+                return $this->totalItems;
+            }
+
+            private function loadDatabase(): void
+            {
+                $this->logger->debug(
+                    sprintf(
+                        'Loading medias from %d/%d',
```
Comment (cg-tw) [inline centreon/src/Core/Media/Infrastructure/Repository/DbReadMediaRepository.php:201]:
```suggestion
                        'Loading media from %d/%d',
```

File: centreon/tests/php/Core/Media/Application/UseCase/MigrateAllMedias/MigrateAllMediasTest.php [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        ->method('isAdmin')
+        ->willReturn(false);
+    ($this->useCase)($this->request, $this->presenter);
+    expect($this->presenter->response)
+        ->toBeInstanceOf(ErrorResponse::class)
+        ->and($this->presenter->response->getMessage())
+        ->toBe(MediaException::operationRequiresAdminUser()->getMessage());
+});
+
+it('should present an MigrateAllMediasResponse when the user is an admin', function (): void{
```
Comment (cg-tw) [inline centreon/tests/php/Core/Media/Application/UseCase/MigrateAllMedias/MigrateAllMediasTest.php:61]:
```suggestion
it('should present a MigrateAllMediasResponse when the user is an admin', function (): void{
```

File: centreon/tests/php/Core/Media/Domain/MediaTest.php [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            'id' => 1,
+            'filename' => 'filename',
+            'directory' => 'directory',
+            'data' => 'data',
+            ...$arguments,
+        ]);
+    };
+});
+
+it('should throw an exception when the id property is lower than 1', function (): void {
```
Comment (cg-tw) [inline centreon/tests/php/Core/Media/Domain/MediaTest.php:42]:
```suggestion
it('should throw an exception when the ID property is lower than 1', function (): void {
```

File: centreon/tests/php/Core/Proxy/Domain/Model/ProxyTest.php [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            'url' => 'localhost',
+            'port' => 0,
+            'login' => 'login',
+            'password' => 'password',
+            ...$arguments,
+        ]);
+    };
+});
+
+it('should throw an exception when the url property is empty', function(): void {
```
Comment (cg-tw) [inline centreon/tests/php/Core/Proxy/Domain/Model/ProxyTest.php:42]:
```suggestion
it('should throw an exception when the URL property is empty', function(): void {
```

File: centreon/tests/php/Core/Proxy/Domain/Model/ProxyTest.php [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+);
+
+it('should throw an exception when the password property is empty', function(): void {
+    ($this->createProxy)(['password' => '   ']);
+})->throws(
+    \Assert\InvalidArgumentException::class,
+    AssertionException::notEmptyString('Proxy:password')->getMessage()
+);
+
+it('should be  an exception when the password property is empty', function(): void {
```
Comment (cg-tw) [inline centreon/tests/php/Core/Proxy/Domain/Model/ProxyTest.php:70]:
```suggestion
it('should throw an exception when the password property is empty', function(): void {
```

**Review by kduret — DISMISSED** [review]
OK for packaging changes

**Review by sc979 — DISMISSED** [review]

**Review by cg-tw — APPROVED** [review]

File: centreon/src/Core/Media/Infrastructure/Repository/ApiWriteMediaRepository.php [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    {
+        $this->timeout = $timeout;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function add(NewMedia $media): int
+    {
+        $apiEndpoint = $this->url . DIRECTORY_SEPARATOR . 'centreon/api/latest/configuration/medias';
```
Comment (a-launois) [inline centreon/src/Core/Media/Infrastructure/Repository/ApiWriteMediaRepository.php:77]: Is it possible for the 'centreon' part to be customized on cloud platforms ?

File: centreon/src/Core/Media/Domain/Model/NewMedia.php [hunk 12]
```diff
@@ -72,4 +74,23 @@ public function getComment(): ?string
...
+     * @throws AssertionFailedException
+     *
+     * @return NewMedia
+     */
+    public static function createFromMedia(Media $media): self
+    {
+        $newMedia = new self(
+            $media->getFilename(),
+            $media->getDirectory(),
+            $media->getData() ?? 'fake_data',
```
Comment (jeremyjaouen) [inline centreon/src/Core/Media/Domain/Model/NewMedia.php:90]:
```suggestion
            $media->getData() ?? '',
```
```suggestion
            $media->getData() ?? '',
```

**Review by jeremyjaouen — APPROVED** [review]

File: centreon/src/Core/Media/Domain/Model/NewMedia.php (hunk 12 above)
Comment (callapa) [reply centreon/src/Core/Media/Domain/Model/NewMedia.php:90]: The property data can not be empty

**Review by kduret — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

**Pull request merged**